### ISSUE TITLE
TestWidgetsFlutterBinding.ensureInitialized() messes up with HTTP requests 

### ISSUE THREAD

**Issue by anuragvohraec** [flutter/flutter#48050]
Hi All,

This is my flutter version:
```
Flutter 1.13.3-pre.23 • channel master • https://github.com/flutter/flutter.git
Tools • Dart 2.8.0 (build 2.8.0-dev.0.0 45db297095)
```

Here is my test case, which basicually connects to a NOSQL db [Couch DB] and try and create a database.
```
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart';

main() async{
  //TestWidgetsFlutterBinding.ensureInitialized();
  test("Test1", () async{
    var createADB = await put("http://0.0.0.0:5984/targetdb");
    expect(createADB.statusCode, 201);

    var deleteADB = await delete("http://0.0.0.0:5984/targetdb");
    expect(deleteADB.statusCode, 200);
  });
}
```

This code works perfectly fine, however I have another lengthy test case where I use dart ffi, where it complains to use : "TestWidgetsFlutterBinding.ensureInitialized()".

So when we uncomment the line `TestWidgetsFlutterBinding.ensureInitialized();` in above code [which is similar to my lengthy test case] it does not works and the status code returned is alwasy 400.

**Comment by mkustermann** [comment]
> however I have another lengthy test case where I use dart ffi

It seems unrelated to `dart:ffi`

The implementation of `TestWidgetsFlutterBinding.ensureInitialized()` seems to install a mock http client in [package:flutter_test/src/_binding_io.dart](https://github.com/flutter/flutter/blob/master/packages/flutter_test/lib/src/_binding_io.dart#L32) (see also [_http/overrides.dart](https://github.com/dart-lang/sdk/blob/master/sdk/lib/_http/overrides.dart)), which is probably why you get a different response.

You can e.g. use a custom http client:
```dart
import 'package:http/io_client.dart';

main() async {
  TestWidgetsFlutterBinding.ensureInitialized();
  ...
  final client = IOClient();
  await client.put("...");
  ...
}
```

/cc @sortie

**Comment by anuragvohraec** [comment]
> The implementation of TestWidgetsFlutterBinding.ensureInitialized() seems to install a mock http client 

That seems to be the cause. 
Can't use a different HTTP client, as the library I use uses http direclty.
And they seems to be causing the problem.

I have manged to have a work around of not using "TestWidgetsFlutterBinding.ensureInitialized()", which was actually required to load assets while testing.

**Comment by Nishnha** [comment]
I think this issue should be re-opened. 

@anuragvohraec pinpointed that `TestWidgetsFlutterBinding.ensureInitialized()` causes the HTTP override, rather than just any `testWidget` like #35318 suggests. 

There also isn't a good fix other than not using the `ensureInitialized()` function, but it's necessary for some tests. I ran into this issue when testing functions that used the `shared_preferences` plugin and the compiler suggested using the `ensureInitialized()` to make sure the environment was in a ready state. I haven't yet found a work-around.

**Comment by Nishnha** [comment]
Actually, I found a fix from #35318 that worked for me.

After calling `TestWidgetsFlutterBinding.ensureInitialized()`, set `HttpOverrides.global  = null` in `dart:io`.

Example.: 
```dart
import 'dart:io' as io;

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();
  io.HttpOverrides.global  = null;
  # tests here
}
```

**Comment by wanosoft** [comment]
> Actually, I found a fix from #35318 that worked for me.
> 
> After calling `TestWidgetsFlutterBinding.ensureInitialized()`, set `HttpOverrides.global = null` in `dart:io`.
> 
> Example.:
> 
> ```dart
> import 'dart:io' as io;
> 
> void main() {
>   TestWidgetsFlutterBinding.ensureInitialized();
>   io.HttpOverrides.global  = null;
>   # tests here
> }
> ```

It fixed my problem! Thanks 💃

**Comment by jmichaels** [comment]
I just ran into this (as a fairly new Flutter user), there may still be some rough edges to iron out here.

I started trying to use the testing framework, and my first test failed with a Flutter error message saying that I should add `TestWidgetsFlutterBinding.ensureInitialized()` to the main method in my test file before the tests.

I did, and it fixed that error, but then I spent all morning trying to figure out why my HTTP calls were only failing (400 error) from inside a `test()` call.

Having the framework helpfully tell you to add something, which then silently intercepts all of your HTTP calls seems...less than ideal.  (Or it's just a *great* troll lol)
